lib/bap_llvm/llvm_elf_loader.hpp

```diff
@@ -0,0 +1,177 @@
+#ifndef LLVM_ELF_LOADER_HPP
+#define LLVM_ELF_LOADER_HPP
+
+#include <iostream>
+#include <iomanip>
+
+#include <llvm/Object/ELFObjectFile.h>
+
+#include "llvm_error_or.hpp"
+#include "llvm_loader_utils.hpp"
+
+namespace loader {
+namespace elf_loader {
+
+using namespace llvm;
+using namespace llvm::object;
+
+static const std::string elf_declarations =
+    "(declare file-type (name str))"
+    "(declare arch (name str))"
+    "(declare entry-point (addr int))"
+    "(declare program-header (name str) (offset int) (size int))"
+    "(declare virtual-program-header (name str) (addr int) (size int))"
+    "(declare program-header-flags (name str) (load bool) (read bool) (write bool) (execute bool))"
+    "(declare section-header (name str) (addr int) (size int))"
+    "(declare symbol-entry (name str) (addr int) (size int))"
+    "(declare code-entry (name str) (addr int))";
+
+template <typename T>
+void file_header(const ELFObjectFile<T> &obj, data_stream &s) {
+    auto hdr = obj.getELFFile()->getHeader();
+    s << "(entry-point " << hdr->e_entry << ")";
+}
+
+std::string name_of_index(std::size_t i) {
+    std::ostringstream s;
+    s << std::setfill('0') << std::setw(2) << i;
+    return s.str();
+}
+
+template <typename I>
+void program_headers(I begin, I end, data_stream &s) {
+    std::size_t i = 0;
+    for (auto it = begin; it != end; ++it, ++i) {
+        bool ld = (it->p_type == ELF::PT_LOAD);
+        bool r = static_cast<bool>(it->p_flags & ELF::PF_R);
+        bool w = static_cast<bool>(it->p_flags & ELF::PF_W);
+        bool x = static_cast<bool>(it->p_flags & ELF::PF_X);
+        auto off = it->p_offset;
+        auto filesz = it->p_filesz;
+        auto name = name_of_index(i);
+        s << "(program-header "  << name << " " << off << " " << filesz << ")";
```

Comment:
tip: you can ask `data_stream`'s `<<` operator to add spaces around words automatically. This will make you definitions much shorter.